library/entropy_poll.c

```diff
@@ -105,59 +110,64 @@ static int getrandom_wrapper( void *buf, size_t buflen, unsigned int flags )
     return( syscall( SYS_getrandom, buf, buflen, flags ) );
 }
 
-#include <sys/utsname.h>
-/* Check if version is at least 3.17.0 */
-static int check_version_3_17_plus( void )
-{
-    int minor;
-    struct utsname un;
-    const char *ver;
-
-    /* Get version information */
-    uname(&un);
-    ver = un.release;
+/* This value is copied from "linux/random.h" to avoid depending on
+ * the presence of this header at build time. Since it is part of the
+ * Linux kernel ABI, it won't change. */
+#define GRND_NONBLOCK 0x0001
 
-    /* Check major version; assume a single digit */
-    if( ver[0] < '3' || ver[0] > '9' || ver [1] != '.' )
-        return( -1 );
+/* Whether the getrandom system call is available at runtime. This will
+ * be tested during program startup in mbedtls_platform_entropy_init(). */
+static int has_getrandom = -1;
+#endif /* defined(SYS_getrandom) */
+#endif /* __linux__ */
 
-    if( ver[0] - '0' > 3 )
-        return( 0 );
+#include <stdio.h>
 
-    /* Ok, so now we know major == 3, check minor.
-     * Assume 1 or 2 digits. */
-    if( ver[2] < '0' || ver[2] > '9' )
-        return( -1 );
+static FILE *urandom_file = NULL;
 
-    minor = ver[2] - '0';
+int mbedtls_platform_entropy_init( void )
+{
+#if defined(HAVE_GETRANDOM)
+    /* Check if the getrandom system call is present. On Linux kernels
+     * older than 3.17, the system call is not present so this returns
+     * ENOSYS. Pass the GRND_NONBLOCK so that this doesn't block even
+     * if the entropy pool is not initialized yet. */
+    has_getrandom = ( getrandom_wrapper( NULL, 0, GRND_NONBLOCK ) == 0 );
+    if( has_getrandom )
+        return( 0);
+#endif /* !defined(HAVE_GETRANDOM) */
 
-    if( ver[3] >= '0' && ver[3] <= '9' )
-        minor = 10 * minor + ver[3] - '0';
-    else if( ver [3] != '.' )
-        return( -1 );
+    if( urandom_file != NULL )
+        fclose( urandom_file );
 
-    if( minor < 17 )
-        return( -1 );
+    urandom_file = fopen( "/dev/urandom", "rb" );
+    if( urandom_file == NULL )
+        return( MBEDTLS_ERR_ENTROPY_SOURCE_FAILED );
 
     return( 0 );
 }
-static int has_getrandom = -1;
-#endif /* SYS_getrandom */
-#endif /* __linux__ */
 
-#include <stdio.h>
+#if defined(HAVE_GETRANDOM)
```

Comment:
This structure of conditionally defining macros is a bit unusual. Can I ask why this was chosen over the more conventional use of conditionally defining different if statements below?

Reply:
I started to write as
```
#if defined(HAVE_GETRANDOM)
    if( has_getrandom == -1 || ( has_getrandom == 0 && urandom_file == NULL ) )
#else
    if( urandom_file == NULL )
#endif
    { ... }
```
but I found the imbrication of build/run-time conditionals confusing.

I don't see what's unusual about defining macros conditionally. The need-init macros are paired with the init functions. There's a single definition for the Unix init function because it has a large amount of shared code. Should `need_init()` be a function with the `#ifdef` inside its body?

Reply:
Am fine with what you did. Just didn't match what I expected.